### ISSUE TITLE
Decide on API for including third party JS libraries

### ISSUE THREAD

**Issue by mspae** [jsdrupal/drupal-admin-ui#476]
Contrib modules should be able to build their own backend config pages in addition to extending existing config pages. They should be able to define their dependencies so they can be used as (modern) JS developers are used to with import or require statements.

However, we don’t want to add a dependency for node and this limits our ability to use modern build tools somewhat. Contrib module developers can be expected to have a node installation in order to build the JS bundles for their module but we want to have the ability to use third-party libraries that can be reused by other modules.

The following solutions are a possibility (please correct and add to this list):
 
1. **Drupal does come with a dependency to node** This is by far the most painless for development at this point, however, it does mean that millions of site maintainers will need to install/or enable node on their installations. This conflicts with our goal of implementing a decoupled backend with minimal friction from the community.
2. **Provide a remote service for bundling of assets** As this is quite a performance intensive task this wouldn’t be trivial (in addition to rate limiting and security aspects)
3. **"Rebuild node in PHP"** or at least provide an intelligent bundling script for JS - This is not really feasible
4. **Modules define their third-party dependencies**, Drupal makes sure they are exposed as global properties on the correct pages and the build (probably webpack) config of the module takes care that imports to that lib are resolved in this way. (See [the webpack external feature documention](https://webpack.js.org/configuration/externals/) – It would make sense to use [the Libraries API](https://www.drupal.org/docs/8/modules/libraries-api-8x) for this

**Comment by dawehner** [comment]
Instead of deciding for an API first u would suggest to see first how/whether it is possible to not require node / skip a build step. Once we know whether it is possible or not, we can much better talk about an api

**Comment by mspae** [comment]
So you're saying: Let's investigate if it is at all possible before deciding on an API? If so then I totally agree. The title sounds maybe too specific. What I mean is: Find out what options we have, discuss, present, document and then at some point decide with which we go.

**Comment by dawehner** [comment]
@mspae Yeah indeed

Do you want to explore the space of not requiring node? I know some other members of the admin UI initiative would like to experiment with the assumption of having node installed, and then see what level of extension points we can add.

**Comment by mspae** [comment]
Yes, will do 👍 and share my results here.

**Comment by mspae** [comment]
### Using composer to install NPM packages, use Libraries API to require them, include them with Libraries API or dynamically with ES6 imports (using the Libraries API to expose their file paths)

Dependencies are resolved with the Drupal libraries API (script tags in order of requirement) or dynamically (see step 5.) – The bundling everything together approach is replaced by parallel loading capabilities of the HTTP2 spec. (Need to make sure Drupal supports this efficiently) – This enables us to use the browser cache efficiently and reduce the amount of bundling/chunking required.

NodeJS is only required for contrib developers, not for site maintainers.

1.  Contrib modules define their third-party dependencies in `composer` with [npm-asset](https://www.google.com/url?q=https://asset-packagist.org/site/about&sa=D&ust=1537181083753000), `composer` is configured so these are put into `web/sites/all/assets/vendor/{$name}` (Could also instead use [foxy](https://www.google.com/url?q=https://github.com/fxpio/foxy&sa=D&ust=1537181083754000) to directly use the contrib module `package.json`)
2.  The Modules also define its dependencies by using the `library_dependencies` key in the theme `info.yml` file (See [the docs for the libraries API](https://www.google.com/url?q=https://www.drupal.org/docs/8/modules/libraries-api-8x/using-libraries-api-3x-as-a-module-developer&sa=D&ust=1537181083754000))
3.  There is a [library definition](https://www.google.com/url?q=https://www.drupal.org/docs/8/modules/libraries-api-8x/library-definition-structure&sa=D&ust=1537181083755000) for each NPM module that is required so the correct file path with the library directory is available
4.  The mapping of library name to correct file path (of UMD bundle(s)) is exposed on the website in JS (probably in `drupalSettings`) (This data is provided by the Libraries API through the library definition files – it just needs to be injected into the JS)
5.  Including JS dependencies can happen in one of two ways:

    1.  Non-dynamic using the built-in library functionality (e.g. `attach_library` etc.) for not fully decoupled admin pages (there is still a normal page load/Drupal is handling the routing)
    2.  Dynamic using a utility library which is used by all other contrib modules. Under the hood it can use [ES6 dynamic imports](https://www.google.com/url?q=https://webpack.js.org/api/module-methods/%23import-&sa=D&ust=1537181083755000) (Currently [only Chrome and Safari support this](https://www.google.com/url?q=https://caniuse.com/%23feat%3Des6-module-dynamic-import&sa=D&ust=1537181083756000), but it is an official spec and will hopefully be widely available by the time this becomes relevant) or any other form of dynamic dependency resolution – It will use the exposed library mapping (see step 4.) to properly map library names with actual files.

6.  Transpiling of JS files happens (`.ejs » .js` file naming pattern) on a per file level (no bundling) – This means only individual files have to be processed.

**Comment by mglaman** [comment]
> (Need to make sure Drupal supports this efficiently)

:D https://www.drupal.org/project/http2_server_push Wim Leers is on it (been on it.)

> with npm-asset, composer is configured so these are put into web/sites/all/assets/vendor/{$name}

Uber nit, but the current practice is `/libraries` of the Drupal root.

> The Modules also define their dependencies by using the library_dependencies key in the theme info.yml file (See the docs for the libraries API)

Could we use core's libraries.yml instead of the Libraries module? Perhaps by adding some additional flag on our own? Like if the license key is provided. See http://cgit.drupalcode.org/dropzonejs/tree/dropzonejs.libraries.yml

I looked at \Drupal\Core\Asset\LibraryDiscovery. It doesn't seem very easy out of the box. We could use `library.discovery.collector`, but `CacheCollectorInterface` does not support a "get all the things."

I don't know if Libraries API gains us much more at a cursory overview. The core library system gives us version definitions and dependencies.

> There is a library definition for each NPM module that is required so the correct file path with the library directory is available

Per above, the core API should let us do that.

> The mapping of library name to correct file path (of UMD bundle(s)) is exposed on the website in JS (probably in drupalSettings) (This data is provided by the Libraries API through the library definition files – it just needs to be injected into the JS)

What if we could get libraries to tag

```
umd: true
```

We could then run our own LibraryManager which discovers all of the system's libraries and finds ones flagged UMD. s a page declares their usage we know to embed them.

Then we could build something which pumps the settings into the page. If we used the normal attached process the actual JS files would be loaded. And then we could build something else to check those and pump it into drupalSettings. Not sure.

We couldn't use a different #attached because it'd through an exception

```
    // Send a message back if the render array has unsupported #attached types.
    $unsupported_types = array_diff(
      array_keys($attached),
      ['html_head', 'feed', 'html_head_link', 'http_header', 'library', 'html_response_attachment_placeholders', 'placeholders', 'drupalSettings']
    );
    if (!empty($unsupported_types)) {
      throw new \LogicException(sprintf('You are not allowed to use %s in #attached.', implode(', ', $unsupported_types)));
    }
```

**Comment by mattgrill** [comment]
@lauriii 

Explore how the Drupal module system can inject changes into the React Admin UI, and after that, explore some specific APIs, such as Field Widgets, and provide prototypes for those.

For more details, see:
docs.google.com/document/d/1U1O4VK0CKbz--aPD5o4WvbQJ6Ss0_T8EX3xntbGdwVo/edit#

**Comment by dawehner** [comment]
# Experiment: Dynamically include widgets without extra build step

https://github.com/dawehner/drupal-admin-ui/tree/experiment-bundliing

Given the assumption we can’t have an extra build step when you install a module we experimented with the following idea

- Modules specify their components in the `.admin_ui.components.yml` file, see above
- Module authors will build their component JS files using Webpack (probably via a helper tool from admin UI, which we will provide)
- We provide an endpoint which points to built JS files

```
{
  widgets: {
    options_buttons: {
      component:
        "http://127.0.0.1:8888/modules/contrib/drupal-admin-ui-support/modules/admin_ui_widget_example/js/build/options_buttons_widget.js",
      multiple: true
    }
  }
}
```
- On runtime the app fetches the the available widgets (widgets from installed modules) using the endpoint above
- The node form loads the JS files using require.js and returns the component
- The form behaves as expected afterwards


**TL;DR Installing a module adds a custom widget without an additional build step.**

## Details

- We tried out the following module loads
- Systemjs
- Script tag injection
- https://github.com/tes/async-define
- Amd vs. umd

We got it  running using UMD + requirejs. This is an implementation detail and we’re more than happy to improve that.

## Problems

- Sharing dependencies is hard
- Requirejs requires us to specify every single material UI component in the LoadComponent file
- Sourcemaps don’t work (yet)

**Comment by larowlan** [comment]
:muscle:

**Comment by dawehner** [comment]
We've added a PR in https://github.com/jsdrupal/drupal-admin-ui/pull/517

**Comment by lauriii** [comment]
> Sharing dependencies is hard

I would make an argument that only a handful of modules depend on external dependencies for creating their admin UI.

> Requirejs requires us to specify every single material UI component in the LoadComponent file

Why do we have to do that?

I'm curious if you looked into the alternative approach of requiring build step for installing modules. Do we know what are the main benefits of that approach and should we still be looking into that?

**Comment by dawehner** [comment]
> I would make an argument that only a handful of modules depend on external dependencies for creating their admin UI.

Well that's true, but we need to make sure that we don't ship with several versions of react and react-dom, given they are incompatible.

> I'm curious if you looked into the alternative approach of requiring build step for installing modules. Do we know what are the main benefits of that approach and should we still be looking into that?

We haven't had a chance to look into that one, especially given that product seems really disappointed by us wanting to introduce this direction. I personally think adding a build step at ```composer require``` time, would result into an overall better user experience, and well, that's maybe indeed a decision we can hopefully make for Drupal 9.
